depends: Eliminate hard dependency on Ubuntu-ABI specific clang

dongcarl opened the pull request:
This is based on:
1. #16392
2. My fixes for #16392: https://github.com/fanquake/bitcoin/compare/329af88c548ff4afe6fed7fdbad3d39f62044048...dongcarl:2019-10-fixing-fanquake-osx-bump

We previously just downloaded a clang binary from llvm.org. However, I think clang is mature and mainstream enough for us to use the system one. The Gitian descriptors have also been changed accordingly (and work).

-----

MarcoFalke if the Gitian Builds fail for this one, it's because:

We **DO** need to fix our docs for generating macOS SDKs... Previously we just tarred up `MacOSX10.14.sdk`, but we also need to copy over the headers as done here: https://github.com/tpoechtrager/osxcross/blob/d3392f4eae78f3fa3f1fd065fa423f2712825102/tools/gen_sdk_package.sh#L183-L188

In fact, we should probably just subtree tpoechtrager's SDK generation scripts.

fanquake committed:
depends: clang 6.0.1

This also removes some now-unnecessary cctools hacks

Co-Authored-By: Cory Fields <cory-nospam-@coryfields.com>

fanquake committed:
build: set minimum supported macOS to 10.12

fanquake committed:
build: use macOS 10.14 SDK

fanquake committed:
depends: native_cctools 921, ld64 409.12, libtapi 1000.10.8

This also removes the obsolete mlinker-version option

Co-Authored-By: Cory Fields <cory-nospam-@coryfields.com>

dongcarl committed:
depends: Eliminate dependency on ABI-specific clang

For the longest time we have downloaded an Ubuntu ABI-specific clang and
just used it. No longer will we be subject to the tyranny of Ubuntu's
ABI!

dongcarl committed:
scripted-diff: depends: Decouple toolchain + binutils

We manually put the toolchain and binutils into the same path in
native_cctools, this is unintuitive. Since we effectively decouple our
toolchain and binutils in this patchset, the make variables should
reflect this as well.

The $(toolchain_path) prefix is removed in the next commit. They are
separate so I can do this scripted-diff.

-BEGIN VERIFY SCRIPT-
find depends -type f -print0 | xargs -0 sed -i "s/_native_toolchain/_native_binutils/g"
find depends -type f -print0 | xargs -0 sed -i "s/toolchain_path/binutils_path/g"
-END VERIFY SCRIPT-

dongcarl committed:
depends: No more binutils_path prefix for CC in site

If we do end up building a compiler inside depends in the future, we can
just add a separate compiler_path prefix and use it here.

See the previous commit for more details.

laanwj commented:
> We DO need to fix our docs for generating macOS SDKs... Previously we just tarred up MacOSX10.14.sdk, but we also need to copy over the headers as done here

If you need people to re-generate the MacOSX SDK, please document it well and make sure there's a way that people without Macs can do it (I don't have access to a Mac anymore, for one). Might want to combine it with the version upgrade.

dongcarl committed:
contrib: macdeploy: Minor README changes.

dongcarl committed:
contrib: macdeploy: Correctly generate macOS SDK

Previously, we did not include the libc++ headers in our SDK creation,
now we can use tpoechtrager's script to correctly do that. The
gen_sdk_package.sh is taken from
https://github.com/tpoechtrager/osxcross/blob/d3392f4eae78f3fa3f1fd065fa423f2712825102/tools/gen_sdk_package.sh

We also document this change.

dongcarl commented:
Updated documentation in 1250d228e14672e4c2e0c83a6548929749c64ce0 and a0fefee822f2bb3f8bcd7ba6a08eb6df9b5c5a35, would like some advice on whether it's alright to just copy the script over from https://github.com/tpoechtrager/osxcross/blob/d3392f4eae78f3fa3f1fd065fa423f2712825102/tools/gen_sdk_package.sh in a0fefee822f2bb3f8bcd7ba6a08eb6df9b5c5a35

MarcoFalke commented:
Is it really required to first bump clang to a more recent version and then in the next commit remove the hardcoded clang? Wouldn't it be more straightforward to just remove the hardcoded clang?
